### ISSUE TITLE
How can I authenticate using Workload identity federation 

### ISSUE THREAD

**Issue by MrPhantomT** [googleapis/google-auth-library-ruby#354]
Hello, I'm trying to authenticate access to Google Drive from my AWS EC2 instance using [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation), but unfortunately can't figure how to do it. 

Is that library support such authentication? Any plans to support it? 

Thanks.

**Comment by quartzmo** [comment]
@MrPhantomT In your EC2 instance, will you obtain a short-lived OAuth 2.0 access token as described in [Workload Identity Federation - Service account impersonation](https://cloud.google.com/iam/docs/workload-identity-federation#impersonation)? If so, I recently added minimal support for using a token that is obtained externally in `v1.1.0`. Usage example in shown in issue #346. Will that work for you? If not, we can investigate how to support the entire flow.

**Comment by MrPhantomT** [comment]
@quartzmo No, I don't have any short-lived token. 

Basically I have tried to follow that [guide](https://cloud.google.com/iam/docs/configuring-workload-identity-federation) trying to allow usage of Google Drive API from AWS EC2 instance that has specific role assigned without obtaining. 

Thanks.

**Comment by quartzmo** [comment]
Is this the relevant part of the guide?

> To allow the use of AWS security credentials, you must configure the workload identity pool to trust your AWS account. Security credentials tokens issued for this AWS account are then recognized by workload identity federation, and you can use the tokens to obtain short-lived service account credentials.

**Comment by MrPhantomT** [comment]
Yes, but in implementation of that library for other languages like python, node it's looks like all steps(parsing of configuration.json file, obtaining of AWS security token) done inside.

**Comment by quartzmo** [comment]
If it's not too much trouble, can you link to where its done in Python and Node?

**Comment by MrPhantomT** [comment]
The next section in the documentation of Node library describes usage: [Using External Identities](https://github.com/googleapis/google-auth-library-nodejs#using-external-identities)
I have tracked code up to the next [file](https://github.com/googleapis/google-auth-library-nodejs/blob/5ed910513451c82e2551777a3e2212964799ef8e/src/auth/awsclient.ts#L102) where all obtaining logic implemented. 

I have a little bit troubles with python but I guess this [file](https://github.com/googleapis/google-auth-library-python/blob/main/google/auth/aws.py#L73) implement all logic and I guess [this part](https://googleapis.dev/python/google-auth/latest/reference/google.auth.aws.html#module-google.auth.aws) of documentation cover it.

**Comment by quartzmo** [comment]
Thank you for these links, that is very helpful. It sounds like you would like to see equivalent feature support in this library?

**Comment by MrPhantomT** [comment]
Yes, it would be great.

**Comment by hermanbanken** [comment]
If we use https://github.com/google-github-actions/auth today with this library (via Fastlane) we get an error:

```
missing client_email
```
which is thrown here:
https://github.com/googleapis/google-auth-library-ruby/blob/9d7051cff4d5e191a5d6756a068e8be539934f0d/lib/googleauth/json_key_reader.rb#L28

The JSON Key generated by https://github.com/google-github-actions/auth has this format:

```json
{
  "type": "external_account",
  "audience": "//iam.googleapis.com/projects/1234567890/locations/global/workloadIdentityPools/default/providers/github",
  "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
  "token_url": "https://sts.googleapis.com/v1/token",
  "service_account_impersonation_url": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/sa-name@project-name.iam.gserviceaccount.com:generateAccessToken",
  "credential_source":
    {
      "url": "https://pipelines.actions.githubusercontent.com/some-long-path/idtoken?api-version=2.0&audience=https%3A%2F%2Fiam.googleapis.com%2Fprojects%2F1234567890%2Flocations%2Fglobal%2FworkloadIdentityPools%2Fdefault%2Fproviders%2Fgithub",
      "headers": { "Authorization": "***" },
      "format": { "type": "json", "subject_token_field_name": "value" },
    },
}
```

So no client_email is present. I think this library should add support for this format, just like all the other Google Auth libraries, which gained support in begin of last year:
- https://github.com/googleapis/google-auth-library-nodejs/pull/1131
- https://github.com/googleapis/google-auth-library-python/pull/698
- https://github.com/googleapis/google-auth-library-java/pull/547

**Comment by hermanbanken** [comment]
Note that libraries like https://github.com/google-github-actions/auth can also generate the OAuth Access Token, so in principal there is no need to understand the `external_account` JSON Key format for most use cases, as long as a pre-populated OAuth  Access Token could be used.

**Comment by hermanbanken** [comment]
@quartzmo do you know if this is a topic that this library is accepting contributions for?

(I'm not comfortable with Ruby, but others in this thread might be more equipped to contribute. Just want to move this thread/feature forward.)

**Comment by rbclark** [comment]
@quartzmo I am quite comfortable with Ruby and would be willing to take a stab at this if it is something the library would entertain merging.

**Comment by corytheboyd-smartsheet** [comment]
Just throwing in my "this would be really really _really_ nice to have in the Ruby SDK". Workload identity federation is going to be huge for us, as we are very multi-cloud but primary running services out of GCP. Huge thanks to @rbclark for taking one for the team running the PR 😄

**Comment by shivgautam** [comment]
Hi, The team is working on it and will post an update soon. We don't have an ETA but it is being actively worked upon and should be ready soon. Thank you all for your patience.

**Comment by kyokyokyo0217** [comment]
Hello, I also have some questions about using Workload Identity Federation with Service Account Impersonation.

I am trying to access Cloud Run on GCP from an EC2 instance on AWS. According to [this](https://cloud.google.com/run/docs/authenticating/service-to-service#use_workload_identity_federation_from_outside), we have to call generateIdToken instead of generateAccessToken to generate a short-lived token.

Here are my questions:

•   In cases like the above, can I use  `Google::Auth::ExternalAccount::Credentials` module?

•   If so, assuming to add a field like `service_account_impersonation_url` and set value like`https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/XX@XX.iam.gserviceaccount.com:generateIdToken` to the configuration file for impersonation? It seemed that the make_creds method only considered json_key_io and scope [here](https://github.com/googleapis/google-auth-library-ruby/blob/main/lib/googleauth/external_account.rb#L38) .

Thanks!

**Comment by aandreassa** [comment]
Workload Identity Federation is fully supported in `google-auth-library-ruby`:

- **AWS Workload Identity:** Added in [PR #418](https://github.com/googleapis/google-auth-library-ruby/pull/418).
- **Identity Pool Credentials (OIDC / EKS):** Added in [PR #433](https://github.com/googleapis/google-auth-library-ruby/pull/433).
- **Pluggable Executable Credentials:** Added in [PR #437](https://github.com/googleapis/google-auth-library-ruby/pull/437).

---

@kyocharoo Regarding your question:
Yes, `Google::Auth::ExternalAccount::Credentials` supports Service Account Impersonation. The `"service_account_impersonation_url"` is specified directly inside the JSON configuration file passed via `json_key_io`. `make_creds` reads that field from the JSON file and handles the impersonation token exchange automatically.

For endpoints requiring OIDC ID tokens (like Cloud Run), you can generate ID tokens using `Google::Auth::IDTokens` or setting `target_audience`.

Closing as resolved!
